[RM Ch 3.5] Network semantics

qabpea opened the pull request:
This is the current state of this chapter, it needs a bit more work though

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
 
+
+<a name="3.5.7"></a>
+ ### 3.5.7 Kubernetes Networking Semantics
```
@qabpea Does this section belong with RA2 [3.2.2 Container Networking Services](https://github.com/cntt-n/CNTT/blob/master/doc/ref_arch/kubernetes/chapters/chapter03.md#322-container-networking-services?)

qabpea replied:
Probably, we should split this once the first version/revision is done

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+<a name="3.5.7.1"></a>
+#### 3.5.7.1 Built in Kubernetes network functionality 
+Kubernetes currently only allows for one network, the *cluster* network and one network attachment for each pod. All pods and containers have an *eth0* interface, this interface is created by kubernetes at pod cration and attached to the *cluster* network. All communication to and from the pod is done using this interface. To only allow for one interface in a pod removes the need for traditional  ...
```
convert the "," after "All pods and containers have an *eth0* interface" either to a period (".") or a semicolon (";").

CsatariGergely replied:
nit: cration -> creation

qabpea replied:
done

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+<a name="3.5.7"></a>
+ ### 3.5.7 Kubernetes Networking Semantics
+The support for traditional network orchestration is more or less non existing in Kubernetes proper. Kubernetes is foremost a Platform as a Service (PaaS) environment and not an Infrastrutcure as a Service (Iaas) infrastucture component. There is no orchetration API like Neutron and no way to instantiate network services such as L3aaS and LBaaS like you can do in Openstack.
```
The sentence "Kubernetes is foremost a Platform as a Service (PaaS) environment and not an Infrastrutcure as a Service (Iaas) infrastucture component." kind of hangs -- expecting something else to follow.

Also, typo's in " infrastucture) --> " infrastructure" (2  places)

CsatariGergely replied:
I think Kubernetes is a bit of IaaS and a bit of SaaS, but inperfect in both roles. I would not state clearly that it is SaaS and not IaaS. I agree that network orchestration is missing though.

CsatariGergely replied:
Openstack -> OpenStack

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+#### 3.5.7.1 Built in Kubernetes network functionality 
+Kubernetes currently only allows for one network, the *cluster* network and one network attachment for each pod. All pods and containers have an *eth0* interface, this interface is created by kubernetes at pod cration and attached to the *cluster* network. All communication to and from the pod is done using this interface. To only allow for one interface in a pod removes the need for traditional  ...
+
+The basic semantics of Kubernetes and the information found in manifest defines the connectivity rules and behavior without in principle any references to IP addresses. This has many advantages, it makes it easy to create portable, scalable SW services and network policies for them that are not location aware and therefore can be executed more or less anywhere. 
```
which "manifest"?

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+Kubernetes built in objects
+Pod and workloads | Description
+------------------|------------
+[Pod:](https://kubernetes.io/docs/concepts/workloads/pods/) | Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.
```
What is a "client" in this respect?

qabpea replied:
I replaced that sentence

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+[Pod:](https://kubernetes.io/docs/concepts/workloads/pods/) | Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.
+[ReplicaSet:](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/) | ReplicaSet ensures that a specified number of pod replicas are running at any given time.
+[Deployment:](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) | Deployment enables declarative updates for Pods and ReplicaSets.
+[DaemonSet:](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/) | DaemonSet represents the configuration of a daemon set.
```
Should the description instead say, DaemonSet is used to manage pods or replicated pods

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+[ReplicaSet:](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/) | ReplicaSet ensures that a specified number of pod replicas are running at any given time.
+[Deployment:](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) | Deployment enables declarative updates for Pods and ReplicaSets.
+[DaemonSet:](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/) | DaemonSet represents the configuration of a daemon set.
+[Job:](https://kubernetes.io/docs/concepts/workloads/controllers/job/) | Job represents the configuration of a single job.
```
Should the description of Job be a configuration for periodic and repeatable tasks

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+[EndpointSlices:](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/) | Endpoints and Endpointslices are a collections of endpoints that contains the ip address, v4 and v6, of the pods that represents a service.
+[Network Policy:](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/) | NetworkPolicy defines which network traffic is allowed to ingress and egress from a set of pods.
+
+There is no need to explicitly define internal load balancers, servers pools, service monitors, firewalls and so on. The kubernetes sematics and relation between the different objects defined in the object manifests contains all the information needed.
```
Typo: "sematics" --> "semantics"

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+There is no need to explicitly define internal load balancers, servers pools, service monitors, firewalls and so on. The kubernetes sematics and relation between the different objects defined in the object manifests contains all the information needed.
+
+Example: The manifests for service *my-service* and the*deployment* with the four load balanced pods of type *my-app*
```
" the*deployment*" ---> " the *deployment*"

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
````diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+                                          - containerPort: 123 
+```
+
+This is all that is needed to deploy 4 pods/containers that are fronted by a service that performe sload balancing. The *Deplyment* will ensure that there are always four pods of type *my-app* available. the *deplyment* is responsible for the full lifecycle management of the pods, this includes in service update/upgrade.
````
"performe sload" ---> "performs load"

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+This is all that is needed to deploy 4 pods/containers that are fronted by a service that performe sload balancing. The *Deplyment* will ensure that there are always four pods of type *my-app* available. the *deplyment* is responsible for the full lifecycle management of the pods, this includes in service update/upgrade.
+
+None of this is of much help however when implementing network service functions such as VNFs/CNFs that requires multi networking and network orchestartion
```
"that requires multi networking and network orchestartion"  ---> "that require multi networking and network orchestration."

pgoyal01 commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+<a name="3.5.7.2"></a>
+#### 3.5.7.2 Multi networking and Orchestration
+Kubernets does curently not in it self support multi networks, pod multi network attachments or network orchestration. This is supported by using a chain loading [*Container Network Interface*](https://github.com/containernetworking/cni) plugin such as [Multus:](https://github.com/k8snetworkplumbingwg/multus-cni).
```
Suggest rewrite of "Kubernets does curently not in it self support multi networks, pod multi network attachments or network orchestration." 

"Kubernetes currently does not in it self support multi networks, pod multi network attachments or network orchestration."

CsatariGergely replied:
Multi interface is not done by chainloading of Multus, but by chainloading of different CNI plugins by Multus. I think this sentance should be clarified. The text can simply refer to CNI multiplexers as this term is already introduced in RA2.

CsatariGergely commented:
Can you please fix the title of this pr?

CsatariGergely commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+The support for traditional network orchestration is more or less non existing in Kubernetes proper. Kubernetes is foremost a Platform as a Service (PaaS) environment and not an Infrastrutcure as a Service (Iaas) infrastucture component. There is no orchetration API like Neutron and no way to instantiate network services such as L3aaS and LBaaS like you can do in Openstack.
+
+Kubernetes networking can be divided into two parts, built in network functinality available through the pod's mandatory primarty interface and network functionality avaiable through the pod's optional secondary inerfaces.
+
```
In ideal cases this division would not exist and Kubernetes would handle all interfaces equaly. At the moment the secondary interfaces are not supported by Kubernetes at all.

qabpea replied:
I rewrote part of this, and as for the multi networking, I hope that this will be non issue within 2-3 years

CsatariGergely commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+The basic semantics of Kubernetes and the information found in manifest defines the connectivity rules and behavior without in principle any references to IP addresses. This has many advantages, it makes it easy to create portable, scalable SW services and network policies for them that are not location aware and therefore can be executed more or less anywhere. 
+
+Kubernetes built in objects
```
I'm not sure if we should copy the description of Kubernetes constructs to here. ..

qabpea replied:
OK, put them in an appendix? I am always struck by that most people have no idea of what the K8s constructs are

CsatariGergely commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+
+There is no need to explicitly define internal load balancers, servers pools, service monitors, firewalls and so on. The kubernetes sematics and relation between the different objects defined in the object manifests contains all the information needed.
+
+Example: The manifests for service *my-service* and the*deployment* with the four load balanced pods of type *my-app*
```
I do not think that this level of detail adds any value to Anuket.

qabpea replied:
OK< lets discuss this,

CsatariGergely commented on doc/ref_model/chapters/chapter03.md:
```diff
@@ -539,6 +542,150 @@ SyncE was standardized by the ITU-T, in cooperation with IEEE, as three recommen
+NetworkAttachmentDefinition object | This defines resource object that describes how to attach a pod to a logical or physical network, the annotation name is *"k8s.v1.cni.cncf.io/networks"*
+Network Attachment Selection Annotation | Selects one or more networks that a pod should be attached to.
+
+Example: Define three network attachments and attach the three networks to a pod.
```
I think this is too match detail.

qabpea replied:
OK, lets discuss this, I rather put everything in and then cut instead of the other way arounf

walterkozlowski commented:
@qabpea, can you please address comments from the reviewers? Thanks.

qabpea commented:
I have created a new branch, just to be sure that merging will be easier. I have addressed all review items except @CsatariGergely comments regarding to much details. I prefer to wait with that discussion and take that decision at the same time as the move of parts to RA2

qabpea commented:
The new pull request is #2546
